drivers/clock_control/clock_control_npcx.c

```diff
@@ -0,0 +1,167 @@
+#include <drivers/clock_control.h>
+#include <dt-bindings/clock/npcx_clock.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(clock_control_npcx, LOG_LEVEL_ERR);
+
+/* Driver config */
+struct npcx_pcc_config {
+	/* cdcg controller base address */
+	cdcg_inst_t base_cdcg;
+	/* pmc controller base address */
+	pmc_inst_t base_pmc;
+};
+
+/* Driver convenience defines */
+#define DRV_CONFIG(dev) \
+	((const struct npcx_pcc_config *)(dev)->config_info)
+
+#define HAL_CDCG_INST(dev) \
+	(cdcg_inst_t)(DRV_CONFIG(dev)->base_cdcg)
+
+#define HAL_PMC_INST(dev) \
+	(pmc_inst_t)(DRV_CONFIG(dev)->base_pmc)
+
+/******************************************************************************/
+/* Clock controller local functions */
+static inline int npcx_clock_control_on(struct device *dev,
+					 clock_control_subsys_t sub_system)
+{
+	ARG_UNUSED(dev);
+	struct npcx_clk_cfg *clk_cfg = (struct npcx_clk_cfg *)(sub_system);
+	pmc_inst_t inst_pmc = HAL_PMC_INST(dev);
+
+	/* Clear related PD (Power-Down) bit of module to turn on clock */
+	CLEAR_BIT(NPCX_PWDWN_CTL(inst_pmc, clk_cfg->ctrl), clk_cfg->bit);
+	return 0;
+}
+
+static inline int npcx_clock_control_off(struct device *dev,
+					  clock_control_subsys_t sub_system)
+{
+	ARG_UNUSED(dev);
+	struct npcx_clk_cfg *clk_cfg = (struct npcx_clk_cfg *)(sub_system);
+	pmc_inst_t inst_pmc = HAL_PMC_INST(dev);
+
+	/* Set related PD (Power-Down) bit of module to turn off clock */
+	SET_BIT(NPCX_PWDWN_CTL(inst_pmc, clk_cfg->ctrl), clk_cfg->bit);
+	return 0;
+}
+
+static int npcx_clock_control_get_subsys_rate(struct device *dev,
+					 clock_control_subsys_t sub_system,
+					 uint32_t *rate)
+{
+	ARG_UNUSED(dev);
+	struct npcx_clk_cfg *clk_cfg = (struct npcx_clk_cfg *)(sub_system);
+
+	switch (clk_cfg->bus)
+	{
+	case NPCX_CLOCK_BUS_APB1:
+		*rate = NPCX_APB_CLOCK(1);
+		break;
+	case NPCX_CLOCK_BUS_APB2:
+		*rate = NPCX_APB_CLOCK(2);
+		break;
+	case NPCX_CLOCK_BUS_APB3:
+		*rate = NPCX_APB_CLOCK(3);
+		break;
+	case NPCX_CLOCK_BUS_AHB6:
+		*rate = CORE_CLK/(AHB6DIV + 1);
+		break;
+	case NPCX_CLOCK_BUS_FIU:
+		*rate = CORE_CLK/(FIUDIV + 1);
+		break;
+	case NPCX_CLOCK_BUS_CORE:
+		*rate = CORE_CLK;
+		break;
+	case NPCX_CLOCK_BUS_LFCLK:
+		*rate = LFCLK;
+		break;
+	default:
+		*rate = 0U;
+		/* Invalid parameters */
+		return -EINVAL;
+	};
+
+	return 0;
+}
+
+/******************************************************************************/
+/* Clock controller driver registration */
+static struct clock_control_driver_api npcx_clock_control_api = {
+	.on = npcx_clock_control_on,
+	.off = npcx_clock_control_off,
+	.get_rate = npcx_clock_control_get_subsys_rate,
+};
+
+static int npcx_clock_control_init(struct device *dev)
+{
+	cdcg_inst_t inst_cdcg = HAL_CDCG_INST(dev);
+	pmc_inst_t inst_pmc  = HAL_PMC_INST(dev);
+
+	/*
+	 * Resting the OSC_CLK (even to the same value) will make the clock
+	 * unstable for a little which can affect peripheral communication like
+	 * eSPI. Skip this if not needed (e.g. RW jump)
```

Comment:
What does 'RW jump' mean? Can you expand this?

Reply:
This a note from the ChromiumOS code base, but I think we can drop the "e.g. RW jump" part.

If we use Early Firmware Selection 2 on newer zephyr based boards, then this shouldn't be an issue, but the general note is still helpful.

Reply:
I will remove (e.g. RW jump) later